### PR TITLE
Implement Harmonized API handler #10522

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/AssetHandler.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import com.enonic.xp.web.WebException;
+import com.enonic.xp.web.WebRequest;
+import com.enonic.xp.web.WebResponse;
+
+import static com.google.common.base.Strings.nullToEmpty;
+
+@Component(service = AssetHandler.class, configurationPid = "com.enonic.xp.portal")
+public class AssetHandler
+{
+    private static final Pattern PATTERN = Pattern.compile( "([^/^:]+)(?::([^/]+))?/(.+)" );
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/AssetHandler.java:33]: this pattern looks incorrect. Does not have a start, and disallows `^` in first path element?

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/AttachmentHandler.java [hunk 2]
```diff
@@ -13,19 +13,15 @@
...
-import com.enonic.xp.web.handler.WebHandler;
-import com.enonic.xp.web.handler.WebHandlerChain;
 
-@Component(immediate = true, service = WebHandler.class, configurationPid = "com.enonic.xp.portal")
-public final class AttachmentHandler
-    extends EndpointHandler
+@Component(service = AttachmentHandler.class)
+public class AttachmentHandler
 {
     private static final Pattern PATTERN = Pattern.compile( "([^/]+)/([^/^:]+)(?::([^/]+))?/([^/]+)" );
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/AttachmentHandler.java:26]: see above about Pattern

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/AttachmentHandler.java [hunk 3]
```diff
@@ -40,9 +36,8 @@ public final class AttachmentHandler
     private volatile String contentSecurityPolicySvg;
 
     @Activate
-    public AttachmentHandler( @Reference final ContentService contentService )
+    public AttachmentHandler( final @Reference ContentService contentService )
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/AttachmentHandler.java:39]: annotations first

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/IdentityHandler.java [hunk 4]
```diff
@@ -41,29 +41,33 @@ public class IdentityHandler
 
     private final RedirectChecksumService redirectChecksumService;
 
+
     @Activate
-    public IdentityHandler( @Reference final ContentService contentService,
-                            @Reference final IdProviderControllerService idProviderControllerService,
-                            @Reference final RedirectChecksumService redirectChecksumService )
+    public IdentityHandler( final @Reference ContentService contentService,
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/IdentityHandler.java:46]: annotations first

**Review by rymsha — CHANGES_REQUESTED** [review]

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/ApiDispatcher.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    @Override
+    protected boolean canHandle( final WebRequest webRequest )
+    {
+        final String path = Objects.requireNonNullElse( webRequest.getEndpointPath(), webRequest.getRawPath() );
+        if ( path.startsWith( "/_/" ) || path.startsWith( "/api/" ) )
+        {
+            final String handler = resolveHandler( webRequest );
+            return switch ( handler )
+            {
+                case "attachment", "image" -> IS_GET_HEAD_OPTIONS_METHOD.test( webRequest ) && IS_SITE_BASE.test( webRequest );
```
[the remark this answers is not in the document]
Comment (anatol-sialitski) [reply modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/ApiDispatcher.java:82]: It is processed here https://github.com/enonic/xp/pull/10533/files#diff-b86a226c097cefaf9a5bbade35cd5575b151c02401c11d1a06db90402967feb8R77

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/AttachmentHandler.java [hunk 6]
```diff
@@ -56,26 +58,34 @@ public void activate( final PortalConfig config )
...
         {
             throw WebException.notFound( "Not a valid attachment url pattern" );
         }
 
+        if ( !IS_GET_HEAD_OPTIONS_METHOD.test( webRequest ) )
+        {
+            throw new WebException( HttpStatus.METHOD_NOT_ALLOWED, String.format( "Method %s not allowed", webRequest.getMethod() ) );
+        }
+
+        if ( !IS_SITE_BASE.test( webRequest ) )
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/AttachmentHandler.java:79]: Should be before method check - not just here but in other handlers, too

**Review by rymsha — CHANGES_REQUESTED** [review]

**Pull request merged**